Friends, Likes.

nixanok opened the pull request.

nixanok committed:
<opt>: Отлажена работа с валидацией данных и централизованным отловом исключений.
<ref>: Зарефакторен код. Добавлены классы сервиса и хранилища. Внедрены зависимости между ними.

nixanok committed:
<fitch>: "Добавлено свойство лайки к фильмам. Реализована логика добавления, удаления, получения лайков. Добавлена возможность получать топ популярных фильмов, отсортированных по лайкам."
<fitch>: "Добавлены друзья. Реализована логика добавления, удаления, получения друзей, а так же общих друзей."

nixanok committed:
<opt>:"Код приведен к чекстайлу."

nixanok committed:
<opt>:"Код приведен к чекстайлу."

nixanok committed:
<opt>:"Код приведен к чекстайлу."

AlexeyCheuzov commented on src/main/java/ru/yandex/practicum/filmorate/controller/errorHandler/ErrorResponse.java:
```diff
@@ -0,0 +1,13 @@
+import java.time.LocalDateTime;
+
+@Data
+public final class ErrorResponse {
```
Не совсем логично располагать модель ответа в папке с контроллерами

AlexeyCheuzov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,91 +1,57 @@
 @RestController
+@RequiredArgsConstructor
 @RequestMapping("/films")
 @Slf4j
```
Если тут подключено логирование, то неплохо было бы добавить логи

AlexeyCheuzov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,91 +1,63 @@
 @RestController
+@RequiredArgsConstructor
 @RequestMapping("/users")
 @Slf4j
```
Если тут подключено логирование, то неплохо было бы добавить логи

AlexeyCheuzov commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -19,8 +25,26 @@ public class Film {
     @Positive(message = "Duration should be positive.")
     private final long duration;
+
+    private final Set<Integer> userLikes = new HashSet<>();
```
Необходимо добавить аннотацию jsonIgnore, чтобы данное поле не учитывалось при сериализации/десериализации

AlexeyCheuzov commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -2,24 +2,38 @@
     @PastOrPresent(message = "Birthday should be in the past or present.")
     private final LocalDate birthday;
+
+    private final Set<Integer> friends = new HashSet<>();
```
Необходимо добавить аннотацию jsonIgnore, чтобы данное поле не учитывалось при сериализации/десериализации

AlexeyCheuzov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,91 +1,57 @@
 
-    public static final LocalDate DATE_OF_FIRST_FILM = LocalDate.of(1895, 12, 28);
+    @Autowired
+    private final FilmService filmService;
```
Более правильным вариантом будет использование интерфейсов для обеспечения слабой связанности, то есть можно сделать интерфейс FilmService, и затем его реализацию, например FilmServiceImpl. Это замечание не обязательное, исправить можно при желании

AlexeyCheuzov commented on src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java:
```diff
@@ -1,91 +1,63 @@
-            return new ResponseEntity<>(error, HttpStatus.BAD_REQUEST);
-        }
+    @Autowired
+    private final UserService userService;
```
Аналогичная ситуация с интерфейсом, как и в FilmController

nixanok committed:
<opt>:"Добавлено логирование в контроллеры."
<ref>:"Модель ошибки перемещена в директорию с моделями."
<opt>:"Список друзей у юзера и лайков у фильма теперь игнорируется json."
<opt>:"Добавлены интерфейсы FilmService и UserService для обеспечения слабой связности".

nixanok committed:
<opt>:"Код приведен к чекстайлу."

nixanok committed:
<opt>:"Код приведен к чекстайлу."

nixanok committed:
<opt>:"Код приведен к чекстайлу."
